virttest/env_process.py

```diff
@@ -314,6 +314,14 @@ def check_image(test, params, image_name, vm_process_status=None):
                               " Skip the image check.")
                 check_image_flag = False
 
+    # Save the potential bad image when the test is not passed.
+    # It should before image check.
+    if (params["test_passed"] == "False" and
+            params.get("save_image_on_error", "no") == "yes"):
```

Comment:
Well the "test_passed" should not be necessary. Standard way of handling params in env_process is to simply add support for a feature (in this case to backup image under a different name), define the default value `save_image=no save_image_on_error=yes` and simply use this `save_image` param. When the test fails, env_process refreshes params to use `_on_error` suffix.

When the test passes, `params.get("save_image")` returns the value of "save_image". When the test fails the same call returns the value of "save_image_on_error" (see `postprocess_on_error`).

Reply:
As to the `postprocess_on_error`, I will have a look tomorrow, thanks

Reply:
Yes, it does work.
It has two problems:
1) if someone uses `save_image=yes`, it will keep the "good" images with your suggested implementation. Is this a necessary action when test passes?
2) it will name the saved image as "xxxx.BAD.hash", which is not accurate.

I still prefer my current implantation as it clearly states its logic: it will only save the "bad" images when the test fails.

Reply:
Well I'm always oppose to custom implementation when existing implementation already exists. Having `on_error` suffix without default workflow breaks assumptions and someone might want to use this to store even good images.

As for the name, yes, it's a bit misleading, which is why the name should be IMO changed to `saved` or `save_image` to connect it to the feature.

My main reason for not breaking existing workflows is that workflows tend to evolve/change and adding simlar implementations of existing workflows might break badly. For example if `on_error` get's renamed your implementation would still use `on_error`, which would be confusing. Consistency is the key to maintainability.

Reply:
I can understand your consistency points. Making code consistency is about following some best practices. If it is a naming convention issue, I can change the param name eg. to `save_bad_image`. 

More importantly, I believe there is no actual use case for `save_image = yes`, which will make a copy of `images` named `xxxx.BAD.hash` no mater case is passed or failed. It is redundant and misleading. 

Regarding the default workflow, I can leave it what used to be (virttest/storage.py +396 to 406). But this new utility for save a "bad" image when test result is not pass, I just want to save a "bad" image.

Reply:
Well I think we can agree to disagree. @luckyh @CongLi what are your takes on this matter? For me consistency is the most important thing so I prefer to create standard feature even if it should be (at this point) only used in `on_error`. There already are some features that are only useful `on_error`, but are implemented as generic features. The reason I prefer to do it this way is that there are no surprises when one tries the feature.

Imagine someone in 10 years trying to save image after testing for whatever reason, they might find this feature, try it manually without error and than having to debug why is it not working without test failure.

As for the naming issue, well, I don't think it has to be named `bad`. We don't know whether the image itself is bad, it could be correct, only the test failed. So I'd simply call it `saved` or simply the image name + random hash. If you insisted on `good`/`bad` names, we could create `test_status = good` and `test_status_on_error = bad` and use it to detect pass/failed test (unless something like that already exists, I haven't checked that).

All in all, I am against hardcoding `on_error`, but if others feel differently I'll withdraw.

Reply:
@ldoktor I'm agree that we'd better not hardcode `on_error`, but I can't imagine why user wants to save images even if test passed.

Anyway, if we have to do so, how about introducing another method (e.g. `save_image()`) to `storage.QemuImg` to handle this stuff?

Reply:
Yes, I'd like that approach. It'd be possible to use this function from other locations as well. As for the usage, I also don't think people will massively use it, but people might want that from time to time manually and having it hardcoded might bring confusion when trying to use it. So what do you say? The name could be a hash or time+date or whatever, I don't really care, but it should work in either way.